### PR TITLE
Updating the 2017 csv file the links are in btagSFProducer https://tw…

### PR DESCRIPTION
**Pull request by scarletnorberg**
https://twiki.cern.ch/twiki/bin/viewauth/CMS/BtagRecommendation94X
Update of the 2017 CSV file for btagging

### CODE REVIEW

**Comment by zhenbinwu** [comment]
From that twiki, it stated that the v5 is only updating the  "iterativefit" scale factors, which we don't need to use. So this PR is not useful for us
[1] These new versions only update the "iterativefit" scale factors. The WP scale factors are the same as in the previous version and in the WP only csv file.

**Comment by zhenbinwu** [comment]
But more real question, how do you apply the fastsim btag SF with deepcsv_13TEV_17SL_18_3_2019.csv files that are already in the folder? I thought that was what this PR should be for.

Also, the 2018 has different files in https://twiki.cern.ch/twiki/bin/viewauth/CMS/BtagRecommendation102X#FastSim_correction_factors. We have deepcsv_13TEV_1718SLSame_18_3_2019ExUnc.csv in the folder.

File: python/postprocessing/modules/btv/btagSFProducer.py [hunk 1]
```diff
@@ -5,19 +5,29 @@
 from PhysicsTools.NanoAODTools.postprocessing.framework.datamodel import Collection
 from PhysicsTools.NanoAODTools.postprocessing.framework.eventloop import Module
 
-def is_relevant_syst_for_shape_corr(flavor_btv, syst):
+def is_relevant_syst_for_shape_corr(flavor_btv, syst, self):
```
Comment (zhenbinwu) [inline python/postprocessing/modules/btv/btagSFProducer.py]: self is a special word in python. This is a bad choice of variable name

File: python/postprocessing/modules/btv/btagSFProducer.py [hunk 2]
```diff
@@ -79,23 +90,50 @@ def __init__(self, era, algo = 'csvv2', sfFileName = None, verbose = 0):
                     },
                     'supported_wp' : [ "L", "M", "T", "shape_corr"]
                 },
-                '2017' : {
-                    'inputFileName' : "DeepCSV_94XSF_V4_B_F.csv",
+                '2016FastSim' : {
+                    'inputFileName' : "deepcsv_13TEV_16SL_18_3_2019.csv",
```
Comment (zhenbinwu) [inline python/postprocessing/modules/btv/btagSFProducer.py:93]: Please add source to this file

File: python/postprocessing/modules/btv/btagSFProducer.py [hunk 3]
```diff
@@ -79,23 +90,50 @@ def __init__(self, era, algo = 'csvv2', sfFileName = None, verbose = 0):
...
+                    'measurement_types' : {
+                        0 : "fastsim",  # b
+                        1 : "fastsim",  # c
+                        2 : "fastsim"   # light
```
Comment (zhenbinwu) [inline python/postprocessing/modules/btv/btagSFProducer.py:97]: What are these? This is expected

File: python/postprocessing/modules/btv/btagSFProducer.py [hunk 4]
```diff
@@ -164,8 +202,12 @@ def __init__(self, era, algo = 'csvv2', sfFileName = None, verbose = 0):
         self.branchNames_central_and_systs = {}
         for central_or_syst in self.central_and_systs:
             if central_or_syst == "central":
+                if era.find("FastSim"):
+                    self.branchNames_central_and_systs[central_or_syst] = "Jet_btagSF_FS"
```
Comment (zhenbinwu) [inline python/postprocessing/modules/btv/btagSFProducer.py:204]: How do people use this?

File: python/postprocessing/modules/btv/btagSFProducer.py (hunk 2 above)
Comment (scarletnorberg) [reply python/postprocessing/modules/btv/btagSFProducer.py:93]: You are correct I did not add a twiki page for any of 2016 but here is the total twiki page so you can get to all other pages from here: https://twiki.cern.ch/twiki/bin/view/CMS/BtagRecommendation
Which is on this line:
https://github.com/susy2015/nanoAOD-tools/blob/Stop0l_dev5_Btag/python/postprocessing/modules/btv/btagSFProducer.py#L57
I will make that more clear. Also here is 2016 for all corrections: 
https://twiki.cern.ch/twiki/bin/viewauth/CMS/BtagRecommendation2016Legacy
I can also add that in if you would like.

File: python/postprocessing/modules/btv/btagSFProducer.py (hunk 3 above)
Comment (scarletnorberg) [reply python/postprocessing/modules/btv/btagSFProducer.py:97]: This is from the csv file that you commented on above and look in the second column, FastSim is what is listed. If by comparison you open the fullSim csv file it has in the second column you have: comb, inc, ect.

File: python/postprocessing/modules/btv/btagSFProducer.py (hunk 4 above)
Comment (scarletnorberg) [reply python/postprocessing/modules/btv/btagSFProducer.py:204]: If you look on the NanoSUSY-tools command on the stop0l_postproc file the FastSim module listed below adds in the btagging module which is where the era is added in. Here is the line: 
btagSFProducer(args.era+"FastSim", algo="deepcsv"),
And the lines in the branch:
https://github.com/susy2015/NanoSUSY-tools/blob/dev_v5_FastSim_Btagging/python/processors/Stop0l_postproc.py#L238-L242

File: python/postprocessing/modules/btv/btagSFProducer.py [hunk 5]
```diff
@@ -5,19 +5,29 @@
 from PhysicsTools.NanoAODTools.postprocessing.framework.datamodel import Collection
 from PhysicsTools.NanoAODTools.postprocessing.framework.eventloop import Module
 
-def is_relevant_syst_for_shape_corr(flavor_btv, syst):
+def is_relevant_syst_for_shape_corr(flavor_btv, syst, self):
     """Returns true if a flavor/syst combination is relevant"""
+    era = self.era
     if flavor_btv == 0:
+        if era.find("FastSim"):
```
Comment (pastika) [inline python/postprocessing/modules/btv/btagSFProducer.py]: this check does NOT do what you think it does.  string.find() in python returns -1 if the string is not found.  "-1" evaluates as True!!!.  Please replace this with
```
if "FastSim" in era:
    #blah
```
Also replace it the 2 other places this is used in this function

File: python/postprocessing/modules/btv/btagSFProducer.py (hunk 1 above)
Comment (pastika) [reply python/postprocessing/modules/btv/btagSFProducer.py]: Agreed, you can just pass era directly here, its the only way "self" is uesd

File: python/postprocessing/modules/btv/btagSFProducer.py [hunk 6]
```diff
@@ -164,8 +202,12 @@ def __init__(self, era, algo = 'csvv2', sfFileName = None, verbose = 0):
         self.branchNames_central_and_systs = {}
         for central_or_syst in self.central_and_systs:
             if central_or_syst == "central":
+                if era.find("FastSim"):
+                    self.branchNames_central_and_systs[central_or_syst] = "Jet_btagSF_FS"
                 self.branchNames_central_and_systs[central_or_syst] = "Jet_btagSF"
             else:
+                if era.find("FastSim"):
```
Comment (pastika) [inline python/postprocessing/modules/btv/btagSFProducer.py]: same comment as above

File: python/postprocessing/modules/btv/btagSFProducer.py [hunk 7]
```diff
@@ -262,7 +304,7 @@ def getSFs(self, jet_data, syst, reader,  measurement_type = 'auto', shape_corr
             # evaluate SF
             sf = None
             if shape_corr:
-                if is_relevant_syst_for_shape_corr(flavor_btv, syst):
+                if is_relevant_syst_for_shape_corr(flavor_btv, syst, self):
```
Comment (pastika) [inline python/postprocessing/modules/btv/btagSFProducer.py]: just pass self.era here, so and make appropriate change where mentioned above.

File: python/postprocessing/modules/btv/btagSFProducer.py (hunk 1 above)
Comment (scarletnorberg) [reply python/postprocessing/modules/btv/btagSFProducer.py]: Agreed was already working on making the change.

File: python/postprocessing/modules/btv/btagSFProducer.py [hunk 8]
```diff
@@ -164,8 +201,12 @@ def __init__(self, era, algo = 'csvv2', sfFileName = None, verbose = 0):
         self.branchNames_central_and_systs = {}
         for central_or_syst in self.central_and_systs:
             if central_or_syst == "central":
+                if era.find("FastSim"):
+                    self.branchNames_central_and_systs[central_or_syst] = "Jet_btagSF_FS"
                 self.branchNames_central_and_systs[central_or_syst] = "Jet_btagSF"
             else:
+                if era.find("FastSim"):
+                    self.branchNames_central_and_systs[central_or_syst] = "Jet_btagSF_FS_%s" % central_or_syst
                 self.branchNames_central_and_systs[central_or_syst] = "Jet_btagSF_%s" % central_or_syst
```
Comment (zhenbinwu) [inline python/postprocessing/modules/btv/btagSFProducer.py]: If I understand correctly, this module will be called twice in the new postporcess. btagSFProducer(args.era+"FastSim", algo="deepcsv"), and btagSFProducer(args.era, algo="deepcsv"). Then this line should within the else condiction of  "if era.find("FastSim")", no?

File: python/postprocessing/modules/btv/btagSFProducer.py [hunk 9]
```diff
@@ -164,8 +201,12 @@ def __init__(self, era, algo = 'csvv2', sfFileName = None, verbose = 0):
         self.branchNames_central_and_systs = {}
         for central_or_syst in self.central_and_systs:
             if central_or_syst == "central":
+                if era.find("FastSim"):
```
Comment (zhenbinwu) [inline python/postprocessing/modules/btv/btagSFProducer.py]: This looks problematic usage of find function. find() will return -1 if not find, which will still pass the if statement (using 0 and non0 I believe in python). A simple "FastSim" in era statement as above is more correct.

**Pull request merged**